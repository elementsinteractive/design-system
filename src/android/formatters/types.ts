export type FormatterConfig = {
  template: string
  header: string
  packageName: string
}
