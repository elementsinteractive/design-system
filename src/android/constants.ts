import { join } from 'path'
import { AndroidTheme } from './types'

export const TEMPLATES = (theme: AndroidTheme) => ({
  typographyTemplate: {
    source: join(__dirname, './templates/typographies.ejs'),
    destination: `${theme.typography}/Type.kt`,
    filter: 'isTypography',
    formatter: 'android/compose/typography',
    packageName: theme.typography.packageName,
  },
  spacingsTemplate: {
    source: join(__dirname, './templates/spacings.ejs'),
    destination: `${theme.spacings}/Spacing.kt`,
    filter: 'isSpacing',
    formatter: 'android/spacings',
    packageName: theme.spacings.packageName,
  },
  colorsTemplate: {
    source: join(__dirname, './templates/colors.ejs'),
    destination: `${theme.colors}/Color.kt`,
    filter: 'color',
    formatter: 'android/compose/colors',
    packageName: theme.colors.packageName,
  },
  headerTemplate: join(__dirname, './templates/header.ejs'),
})
