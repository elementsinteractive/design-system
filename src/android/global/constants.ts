import { join } from 'path'

import { AndroidGlobalOptions } from '../types'

export const ColorTypeFilename = 'ColorTokens.kt'

export const SpacingTypeFilename = 'SpacingTokens.kt'

export const GLOBAL_TEMPLATES = (options: AndroidGlobalOptions) => ({
  colorsTypeTemplate: createColorsTypeTemplate(options),
  spacingsTypeTemplate: createSpacingsTypeTemplate(options),
  headerTemplate: join(__dirname, './templates/header.ejs'),
})

const createSpacingsTypeTemplate = (options: AndroidGlobalOptions) => ({
  source: join(__dirname, './templates/spacingTokens.ejs'),
  destination: `${options.destination}/${SpacingTypeFilename}`,
  filter: 'spacingsType',
  formatter: 'android/spacingsType',
  packageName: options.packageName,
})

const createColorsTypeTemplate = (options: AndroidGlobalOptions) => ({
  source: join(__dirname, './templates/colorTokens.ejs'),
  destination: `${options.destination}/${ColorTypeFilename}`,
  filter: 'colorsType',
  formatter: 'android/compose/colorsType',
  packageName: options.packageName,
})
