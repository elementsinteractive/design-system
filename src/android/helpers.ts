import { readFileSync } from 'fs'
import { fileToJSON } from '../common/helpers'
import { AndroidJsonOptions } from './types'

export const readOptionsAsJson = (configPath: string) =>
  fileToJSON(configPath) as AndroidJsonOptions
