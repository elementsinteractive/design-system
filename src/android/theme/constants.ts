import { join } from 'path'
import { AndroidThemeOptions } from '../types'

export const PaletteFilename = 'Palette.kt'
export const SpacingFilename = 'Spacing.kt'
export const TypeFilename = 'Type.kt'
export const ColorTokensFilename = 'ColorTokens.kt'

export const THEME_TEMPLATES = (options: AndroidThemeOptions) => ({
  typographyTemplate: createTypographyTemplate(options),
  spacingsTemplate: createSpacingsTemplate(options),
  paletteTemplate: createPaletteTemplate(options),
  colorTokensTemplate: createColorTokensTemplate(options),
  headerTemplate: join(__dirname, './templates/header.ejs'),
})

const createTypographyTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/typographies.ejs'),
  destination: `${theme.typography.destination}/${TypeFilename}`,
  filter: 'typography',
  formatter: 'android/compose/typography',
  packageName: theme.typography.packageName,
})

const createSpacingsTemplate = (options: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/base/spacings.ejs'),
  destination: `${options.theme.spacings.destination}/${SpacingFilename}`,
  filter: 'spacings',
  formatter: 'android/spacings',
  packageName: options.theme.spacings.packageName,
  sharedPackageName: options.globalPackageName,
})

const createPaletteTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/shared/palette.ejs'),
  destination: `${theme.palette.destination}/${PaletteFilename}`,
  filter: 'color',
  formatter: 'android/compose/palette',
  packageName: theme.palette.packageName,
  themeName: theme.name,
})

const createColorTokensTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/shared/colorTokens.ejs'),
  destination: `${theme.colors.destination}/${ColorTokensFilename}`,
  filter: 'color',
  formatter: 'android/compose/colorTokens',
  packageName: theme.colors.packageName,
  themeName: theme.name,
})
