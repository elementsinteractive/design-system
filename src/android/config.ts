import { AndroidConfig } from '../design.types'
import { TEMPLATES } from './constants'
import { AndroidJsonOptions } from './types'

export const createConfig = (input: string, options: AndroidJsonOptions) => {
  const { theme } = options

  const { colorsTemplate, spacingsTemplate, typographyTemplate } =
    TEMPLATES(theme)

  return {
    source: [input],
    platforms: {
      android: {
        transformGroup: 'tokens-android',
        files: [typographyTemplate, spacingsTemplate, colorsTemplate].map(
          template => ({
            format: template.formatter,
            destination: template.destination,
            filter: template.filter,
          }),
        ),
      },
    },
  }
}
