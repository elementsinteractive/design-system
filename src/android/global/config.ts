import { AndroidGlobalOptions } from '../types'
import { GLOBAL_TEMPLATES } from './constants'

export const createGlobalConfig = (
  input: string,
  options: AndroidGlobalOptions,
) => {
  const { spacingsTypeTemplate, colorsTypeTemplate } = GLOBAL_TEMPLATES(options)

  return {
    source: [input],
    platforms: {
      android: {
        transformGroup: 'tokens-android',
        files: [spacingsTypeTemplate, colorsTypeTemplate].map(template => ({
          format: template.formatter,
          destination: template.destination,
          filter: template.filter,
        })),
      },
    },
  }
}
