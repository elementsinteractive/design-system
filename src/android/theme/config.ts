import { THEME_TEMPLATES } from './constants'
import { AndroidThemeOptions } from '../types'

export const createThemeConfig = (
  input: string,
  options: AndroidThemeOptions,
) => {
  const {
    paletteTemplate,
    spacingsTemplate,
    typographyTemplate,
    colorTokensTemplate,
  } = THEME_TEMPLATES(options)

  return {
    source: [input],
    platforms: {
      android: {
        transformGroup: 'tokens-android',
        files: [
          typographyTemplate,
          spacingsTemplate,
          paletteTemplate,
          colorTokensTemplate,
        ].map(template => ({
          format: template.formatter,
          destination: template.destination,
          filter: template.filter,
        })),
      },
    },
  }
}
