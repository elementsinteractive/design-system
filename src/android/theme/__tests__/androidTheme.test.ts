import StyleDictionaryPackage from 'style-dictionary'
import { clearOutput, fileExists, fileToString } from '../../../common/helpers'
import { createThemeConfig } from '../config'
import { PaletteFilename, SpacingFilename, TypeFilename } from '../constants'
import { readJsonFileAsOptions } from '../../helpers'
import { AndroidThemeOptions } from '../../types'
import { getThemeSD } from '..'

describe('android', () => {
  var config = {
    configPath: 'src/android/theme/__tests__/darkThemeConfig.json',
    input: 'src/android/theme/__tests__/darkThemeTokens.json',
  }

  let StyleDictionaryExtended

  beforeAll(() => {
    const { input, configPath } = config

    StyleDictionaryExtended = getThemeSD(configPath, input)
  })

  beforeEach(() => {
    clearOutput('src/android/theme/__tests__/__output')
  })

  it('can setup with theme specific config', () => {
    StyleDictionaryExtended.buildPlatform('android')

    expect(
      StyleDictionaryExtended.transformGroup['tokens-android'],
    ).toMatchSnapshot()

    const options = readJsonFileAsOptions<AndroidThemeOptions>(
      config.configPath,
    )

    const paletteFile = `${options.theme.palette.destination}/${PaletteFilename}`
    const typeFile = `${options.theme.typography.destination}/${TypeFilename}`

    const files = [paletteFile, typeFile]

    files.forEach(path => {
      console.log('path', path)
      expect(fileExists(path)).toBeTruthy()
      expect(fileToString(path)).toMatchSnapshot()
    })
  })
})
