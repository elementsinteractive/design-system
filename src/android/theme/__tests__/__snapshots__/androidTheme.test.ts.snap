// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`android can setup with theme specific config 1`] = `
Array [
  "name/ti/camel",
]
`;

exports[`android can setup with theme specific config 2`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package org.example.design.base

import androidx.compose.ui.graphics.Color

val primary = Color(0xffff495b)

val background = Color(0xff000000)

val primaryVariant = Color(0xfff4f4f4)

val secondary = Color(0xffffffff)

val secondaryVariant = Color(0xfff4f4f4)

val surface = Color(0xff000000)

val onPrimary = Color(0xffffffff)

val onSecondary = Color(0xff000000)

val onBackground = Color(0xffffffff)

val onSurface = Color(0xffffffff)

val purple = Color(0xff5048da)

val white = Color(0xffffffff)

val pink = Color(0xffff495b)

val black = Color(0xff000000)

val lightGray = Color(0xfff4f4f4)

"
`;

exports[`android can setup with theme specific config 3`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package org.example.design.theme

import androidx.compose.material.Typography
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp

val bodyLarge = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 16.sp,
    letterSpacing = (0.5).sp,
)

val bodyMedium = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 14.sp,
    letterSpacing = (0.2).sp,
)

val bodySmall = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 12.sp,
    letterSpacing = (0.4).sp,
)

val displayLarge = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 57.sp,
    letterSpacing = (-0.2).sp,
)

val displayMedium = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 45.sp,
)

val displaySmall = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 36.sp,
)

val headlineLarge = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 32.sp,
)

val headlineMedium = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 28.sp,
)

val headlineSmall = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 24.sp,
)

val labelLarge = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 14.sp,
    letterSpacing = (0.1).sp,
)

val labelMedium = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 12.sp,
    letterSpacing = (0.5).sp,
)

val labelSmall = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 11.sp,
    letterSpacing = (0.5).sp,
)

val titleLarge = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 22.sp,
)

val titleMedium = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 16.sp,
    letterSpacing = (0.2).sp,
)

val titleSmall = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 14.sp,
    letterSpacing = (0.1).sp,
)

val AppTypography = Typography(
    bodyLarge = bodyLarge,
    bodyMedium = bodyMedium,
    bodySmall = bodySmall,
    displayLarge = displayLarge,
    displayMedium = displayMedium,
    displaySmall = displaySmall,
    headlineLarge = headlineLarge,
    headlineMedium = headlineMedium,
    headlineSmall = headlineSmall,
    labelLarge = labelLarge,
    labelMedium = labelMedium,
    labelSmall = labelSmall,
    titleLarge = titleLarge,
    titleMedium = titleMedium,
    titleSmall = titleSmall,
)"
`;
